### PR TITLE
0094-feel-product-function

### PR DESCRIPTION
**Pull request by StrayAlien**
Not a DMN 1.3 thing, but we had no coverage on this guy.

### CODE REVIEW

**Comment by opatrascoiu** [comment]
I've noticed that the expected value for `product([])` is 0. The standard does not specify the behaviour for this situations, however it specifies the expected value for `sum([])` - **null**. I think we should be consistent.

**Comment by StrayAlien** [comment]
Thanks @opatrascoiu  - good call.  That seems fair enough.  I am happy with that.  Though, fyi, I just checked RH (which I think is the 'reference' here ?) - it assumes zero: https://github.com/kiegroup/drools/blob/master/kie-dmn/kie-dmn-feel/src/main/java/org/kie/dmn/feel/runtime/functions/ProductFunction.java#L39.  And a test for it: https://github.com/kiegroup/drools/blob/master/kie-dmn/kie-dmn-signavio/src/test/java/org/kie/dmn/signavio/feel/runtime/FEELExtendedFunctionsTest.java#L68

**Comment by tarilabs** [comment]
DMN RTF jira https://issues.omg.org/browse/DMN14-141

**Comment by StrayAlien** [comment]
Thanks @tarilabs - to help keep things moving, I have commented out that product([]) test.

**Comment by StrayAlien** [comment]
Another note, @opatrascoiu , @tarilabs - is that it is likely that `1` is actually the correct answer, not `0` or `null`.  Hunting about seems to reveal the 'empty product' of a list of numbers is 1. https://en.wikipedia.org/wiki/Empty_product, https://math.stackexchange.com/questions/110546/what-is-the-product-of-the-empty-set.  I sure am no expert, but it is possible the spec assumes knowledge of this.

**Comment by opatrascoiu** [comment]
@StrayAlien @tarilabs If `product([]) = 1` than `sum([]) = 0`. We might want to change that one as well.

**Comment by tarilabs** [comment]
Good catch @opatrascoiu with https://github.com/dmn-tck/tck/pull/392#issuecomment-788038767

Now I think I recall.
The definition of these function has implication on Decision Table Hit Policy.
It was decided to be consistent with the "SQL representation" during [this DMNv1.2 RTF work](https://issues.omg.org/browse/DMN12-176?focusedCommentId=17495&page=com.atlassian.jira.plugin.system.issuetabpanels:comment-tabpanel#comment-17495).

Sorry, @StrayAlien , I am afraid I was too quick/optimist.
I believe there are serious chances that, even if a Hit Policy Collect-Product does not exists, for consistency reasons we will end up having `product([]) = null`.

**Comment by StrayAlien** [comment]
Hey @tarilabs, @opatrascoiu.  Thanks. I'm fine any which way, as long as we agree and we all do the same ... and we have a test. 👍 .  I'll leave said test commented out for now until RTF provides clarity.

For posterity, here is that DMN 1.2 link without a login.  https://issues.omg.org/issues/DMN12-176?focusedCommentId=17495&page=com.atlassian.jira.plugin.system.issuetabpanels:comment-tabpanel#comment-17495.  (just replace `/browse` with `/issues`)

**Comment by tarilabs** [comment]
> here is that DMN 1.2 link without a login

unfortunately comments are not publicly available, and the same goes for "proposal" jira entries.
I will try to be more diligent and link the public link when the information is indeed publicly available, but in this case the URL I posted is the intended one (linking to a comment).

**Comment by StrayAlien** [comment]
Ahh.  No comments there.  Sorry!  I just did my usual "replace /browse with /issues" to get there.  No offence meant.

File: TestCases/compliance-level-3/0094-feel-product-function/0094-feel-product-function.dmn [hunk 1]
```diff
@@ -0,0 +1,144 @@
...
+             xmlns="https://www.omg.org/spec/DMN/20191111/MODEL/"
+             xmlns:di="http://www.omg.org/spec/DMN/20180521/DI/"
+             xmlns:dmndi="https://www.omg.org/spec/DMN/20191111/DMNDI/"
+             xmlns:dc="http://www.omg.org/spec/DMN/20180521/DC/"
+             xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance">
+
+    <description>FEEL built-in function 'product(list)' in category list functions</description>
+
+    <decision name="decision001" id="_decision001">
+        <description>Tests FEEL expression: 'product([6, 3, 9, 6, 6])' and expects result: '[6] (list)'</description>
```
Comment (tarilabs) [inline TestCases/compliance-level-3/0094-feel-product-function/0094-feel-product-function.dmn:14]:
```suggestion
        <description>Tests FEEL expression: 'product([6, 3, 9, 6, 6])' and expects result: '5832 (number)'</description>
```

File: TestCases/compliance-level-3/0094-feel-product-function/0094-feel-product-function.dmn [hunk 2]
```diff
@@ -0,0 +1,144 @@
...
+        </literalExpression>
+    </decision>
+
+    <decision name="decision004" id="_decision004">
+        <description>Tests FEEL expression: 'product([1,2,null,4])' and expects result: 'null (list)'</description>
+        <question>Result of FEEL expression 'product([1,2,null,4])'?</question>
+        <allowedAnswers>null (list)</allowedAnswers>
+        <variable name="decision004"/>
+        <literalExpression>
+            <text>product()</text>
```
Comment (tarilabs) [inline TestCases/compliance-level-3/0094-feel-product-function/0094-feel-product-function.dmn:59]:
```suggestion
            <text>product([1,2,null,4])</text>
```

File: TestCases/compliance-level-3/0094-feel-product-function/0094-feel-product-function.dmn [hunk 3]
```diff
@@ -0,0 +1,144 @@
...
+        <question>Result of FEEL expression 'product([1,2,"foo",4])'?</question>
+        <allowedAnswers>null (list)</allowedAnswers>
+        <variable name="decision005"/>
+        <literalExpression>
+            <text>product([1,2,"foo",4])</text>
+        </literalExpression>
+    </decision>
+
+    <decision name="decision006" id="_decision006">
+        <description>Tests FEEL expression: 'product([2.5, 1, 2.5, 3])' and expects result: '2.5 (list)'</description>
```
Comment (tarilabs) [inline TestCases/compliance-level-3/0094-feel-product-function/0094-feel-product-function.dmn:74]:
```suggestion
        <description>Tests FEEL expression: 'product([2.5, 1, 2.5, 3])' and expects result: '18.75 (number)'</description>
```

File: TestCases/compliance-level-3/0094-feel-product-function/0094-feel-product-function.dmn [hunk 4]
```diff
@@ -0,0 +1,144 @@
...
+        <question>Result of FEEL expression 'product([2.5, 1, 2.5, 3])'?</question>
+        <allowedAnswers>2.5 (number)</allowedAnswers>
```
Comment (tarilabs) [inline TestCases/compliance-level-3/0094-feel-product-function/0094-feel-product-function.dmn:76]:
```suggestion
        <allowedAnswers>18.75 (number)</allowedAnswers>
```

File: TestCases/compliance-level-3/0094-feel-product-function/0094-feel-product-function.dmn [hunk 5]
```diff
@@ -0,0 +1,144 @@
...
+        <question>Result of FEEL expression 'product([6])'?</question>
+        <allowedAnswers>6 (number)</allowedAnswers>
+        <variable name="decision008_a"/>
+        <literalExpression>
+            <text>product([6])</text>
+        </literalExpression>
+    </decision>
+
+    <decision name="decision011" id="_decision011">
+        <description>Tests FEEL expression: 'product(list:[6, 3, 9, 6, 6])' and expects result: '[6] (list)'</description>
```
Comment (tarilabs) [inline TestCases/compliance-level-3/0094-feel-product-function/0094-feel-product-function.dmn:114]:
```suggestion
        <description>Tests FEEL expression: 'product(list:[6, 3, 9, 6, 6])' and expects result: '5832 (number)'</description>
```

File: TestCases/compliance-level-3/0094-feel-product-function/0094-feel-product-function.dmn [hunk 6]
```diff
@@ -0,0 +1,144 @@
...
+        <question>Result of FEEL expression 'product(list:[6, 3, 9, 6, 6])'?</question>
+        <allowedAnswers>[6] (list)</allowedAnswers>
```
Comment (tarilabs) [inline TestCases/compliance-level-3/0094-feel-product-function/0094-feel-product-function.dmn:116]:
```suggestion
        <allowedAnswers>5832 (number)</allowedAnswers>
```

**Review by tarilabs — COMMENTED** [review]
Thank you!
I'm pretty positive some changes are needed, below.
What do you think?
(you can "approve" and effectively incorporate them directly from this github page)

**Comment by tarilabs** [comment]
By the way, we discussed this at the RTF meeting held 2021-03-23, and the agreement is for `product([])=null`, but I noticed this was indeed elided from the tests; thanks, we can re-include it when DMNv1.4 is out. If you want the change reference is https://issues.omg.org/browse/DMN14-155

**Comment by StrayAlien** [comment]
Thanks @tarilabs - they look good - accepted.  Apols for my sloppy attention to detail there.  Might have been a long day.  And that style of test has a lot of duplications/boilerplate.  Thanks again.

**Comment by StrayAlien** [comment]
Oh, and re the `product([])=null`, yes, it was commented out, but I'd be happy to include it back.  It is not in conflict with the current spec so, as an agreed-upon-at-RTF thing, maybe we just get it back in here.

**Comment by tarilabs** [comment]
no worries at all @StrayAlien , thank you again for this contribution. Yeah, feel free to uncomment `product([])=null` on the basis indeed we do not contradict DMNv1.3 either, and we know this will be an explicit example in DMNv1.4.
That way I can remove the "waiting on RTF" label

**Comment by StrayAlien** [comment]
done

**Comment by tarilabs** [comment]
removing label "Waiting on RTF" per https://github.com/dmn-tck/tck/pull/392#issuecomment-814719019

**Comment by tarilabs** [comment]
TCK meeting of 2021-04-15 agreed, and merging.

**Pull request merged**